### PR TITLE
Convert Upgrader Script to Go binary

### PR DESCRIPTION
**Pull request by rahulbabu95**
*Issue #, if available:*
Upgrader project consumed by eks-a for InPlace upgrades uses an upgrader script written in shell for running the upgrade commands on the node. Ideally, we need it as a Go binary for long term maintainability, testability and general conformance with other parts of the codebase. 

*Description of changes:*
This change converts the upgrader shell script into a Go binary along with unit tests to cover the core business logic. Manually tested by copying over the Go binary to the nodes to test if the corresponding upgrade logic is working as expected when the binary is invoked instead of the shell script. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: projects/aws/upgrader/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package cmd
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/aws/eks-anywhere-build-tooling/tools/version-tracker/pkg/util/logger"
```
Comment (abhinavmpandey08) [inline projects/aws/upgrader/cmd/root.go:10]: Was discussing with @abhay-krishna. Can we move this logger to `eks-anywhere-build-tooling/pkg` and refactor this and version-tracker projects to use that?

File: projects/aws/upgrader/Makefile [hunk 2]
```diff
@@ -12,15 +13,25 @@ EXCLUDE_FROM_UPGRADE_BUILDSPEC=true
...
+SOURCE_PATTERNS=. 
+GO_MOD_PATHS=..
+
+# force binaries to go to non-release branch bin folder
+BINARIES_ARE_RELEASE_BRANCHED=false
+GIT_CHECKOUT_TARGET=main.go
+FAKE_GIT_REPO_TARGET=.git
+REPO_NO_CLONE=true
+HAS_LICENSES=false
+SKIP_CHECKSUM_VALIDATION=true
```
Comment (abhay-krishna) [inline projects/aws/upgrader/Makefile:26]: Let's keep license generation and checksum validation for this project.

File: projects/aws/upgrader/cmd/upgrade_node.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		}
+		return nil
+	},
+}
+
+func init() {
+	var err error
+
+	upgradeCmd.AddCommand(upgradeNodeCmd)
+	upgradeNodeCmd.Flags().String("type", "", "node type flag")
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/cmd/upgrade_node.go:40]: Pass this in as a flag instead of as a string? Do we anticipate significantly more here?

File: projects/aws/upgrader/upgrade/cniplugins_upgrade.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	"context"
+	"fmt"
+
+	"github.com/aws/eks-anywhere-build-tooling/tools/version-tracker/pkg/util/logger"
+)
+
+func (u *Upgrader) CniPluginsUpgrade(ctx context.Context) error {
+	cmpDir, err := u.upgradeComponentsBinDir()
+	if err != nil {
+		return fmt.Errorf("getting upgrade componenets binary directory: %v", err)
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/cniplugins_upgrade.go:13]:
```suggestion
		return fmt.Errorf("getting upgrade components binary directory: %v", err)
```

File: projects/aws/upgrader/upgrade/cniplugins_upgrade.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+		return execError(cpCmd, string(out))
+	}
+
+	out, err = u.ExecCommand(ctx, cniVersionCmd[0], cniVersionCmd[1:]...)
+	if err != nil {
+		return execError(cniVersionCmd, string(out))
+	}
+
+	logger.Info("Cni-Plugins Version on the Node", "Version", string(out))
+	logger.Info("Cni-Plugins upgrade succesful!")
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/cniplugins_upgrade.go:34]: Do we need to camel case this way? Looks off

File: projects/aws/upgrader/upgrade/containerd_upgrade.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/eks-anywhere-build-tooling/tools/version-tracker/pkg/util/logger"
+)
+
+func (u *Upgrader) ContainerdUpgrade(ctx context.Context) error {
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/containerd_upgrade.go:10]: Should we add comments to this business logic to explain what is happening as part of this flow?

File: projects/aws/upgrader/upgrade/cniplugins_upgrade_test.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+	"context"
+	"fmt"
+	"testing"
+)
+
+const (
+	cmpDir = "/foo/binaries"
+)
+
+func TestCniPluginsUpgrade(t *testing.T) {
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/cniplugins_upgrade_test.go:13]: Testing error cases?

File: projects/aws/upgrader/upgrade/kubeadm_upgrade.go [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+	newKubeAdmConfig := fmt.Sprintf("%s/kubeadm-config.yaml", componentsDir)
+
+	getClusterConfigCmd := []string{"kubectl", "get", "cm", "-n", "kube-system", "kubeadm-config", "-ojsonpath='{.data.ClusterConfiguration}'", "--kubeconfig", kubeConfigPath}
+	out, err := u.ExecCommand(ctx, getClusterConfigCmd[0], getClusterConfigCmd[1:]...)
+	if err != nil {
+		return execError(getClusterConfigCmd, string(out))
+	}
+
+	trimmedOut := strings.Trim(string(out), "`'")
+	err = u.WriteFile(kubeAdmConfigBackUp, []byte(trimmedOut), 0o640)
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubeadm_upgrade.go:41]: Should we make the permission a const as well?

File: projects/aws/upgrader/upgrade/kubeadm_upgrade.go [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+	err = u.WriteFile(kubeAdmConf, []byte(conf), 0o640)
+	if err != nil {
+		return fmt.Errorf("writing kubelet config to kubeadm file: %v", err)
+	}
+
+	logger.Info("Appended Kubelet Config to Kubeadm config file")
+	return nil
+}
+
+// Backup and delete coredns configmap. If the CM doesn't exist, kubeadm will skip its upgrade.
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubeadm_upgrade.go:229]: Yea this is good

File: projects/aws/upgrader/upgrade/upgrade_status.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+func (u *Upgrader) LogStatusAndCleanup(ctx context.Context) error {
+	if err := u.logStatus(ctx); err != nil {
+		return fmt.Errorf("logging current status of the node: %v", err)
+	}
+
+	upgCmpDir, err := u.upgradeComponentsDir()
+	if err != nil {
+		return fmt.Errorf("getting upgrade componenets kubernetes binary directory: %v", err)
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/upgrade_status.go:17]:
```suggestion
		return fmt.Errorf("getting upgrade components kubernetes binary directory: %v", err)
```

File: projects/aws/upgrader/upgrade/syscalls.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+package upgrade
+
+import (
+	"context"
+	"io/fs"
+	"os"
+	"os/exec"
+)
+
+type SysCalls interface {
```
Comment (jacobweinstock) [inline projects/aws/upgrader/upgrade/syscalls.go:12]: Why is an interface needed? Are there multiple implementations for this? Also, this interface is too big.

File: projects/aws/upgrader/upgrade/upgrader.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
+package upgrade
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/aws/eks-anywhere-build-tooling/tools/version-tracker/pkg/util/logger"
```
Comment (jacobweinstock) [inline projects/aws/upgrader/upgrade/upgrader.go:8]: What value does this library provide over just using the `logr` package directly? Also, this package has not avoided package state in its implementation. There is no reason for a logger to have package state, it's a solved problem. I recommend not using it and to make all dependencies explicit by passing them into funds and methods.

**Review by jacobweinstock — CHANGES_REQUESTED** [review]
You should have code comments on all exported functions. I highly suggest you add more code comments around why things are being done the way they are, why code exists, etc.

File: projects/aws/upgrader/cmd/upgrade_node.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+	"fmt"
+	"log"
+
+	"github.com/aws/eks-anywhere-build-tooling/projects/aws/upgrader/upgrade"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var upgradeNodeCmd = &cobra.Command{
+	Use:   "node",
```
Comment (jacobweinstock) [inline projects/aws/upgrader/cmd/upgrade_node.go:15]: This naming doesn't seem to follow the pattern of other commands. Kubeadm is not included in the name, only in the description.

File: projects/aws/upgrader/main.go [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+	"os"
+
+	"github.com/aws/eks-anywhere-build-tooling/projects/aws/upgrader/cmd"
+)
+
+func main() {
+	if cmd.Execute() == nil {
+		os.Exit(0)
+	}
+	os.Exit(-1)
```
Comment (jacobweinstock) [inline projects/aws/upgrader/main.go:13]: we are losing the error message here.

File: projects/aws/upgrader/upgrade/cniplugins_upgrade.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
+package upgrade
```
Comment (jacobweinstock) [inline projects/aws/upgrader/upgrade/cniplugins_upgrade.go:1]: Files in the upgrade package with `_upgrade` in their names do not provide any additional value as they are all already in the upgrade package. I suggest removing this `_upgrade`.

File: projects/aws/upgrader/upgrade/upgrader.go [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+)
+
+type Upgrader struct {
```
Comment (jacobweinstock) [inline projects/aws/upgrader/upgrade/upgrader.go:11]: Naming is hard and I recommend the following:

Structs as plain nouns: API, Replica, Object
Interfaces are active nouns: Reader, Writer, JobProcessor
Functions and methods are verbs: Read, Process, Sync 

ref: https://peter.bourgon.org/blog/2019/04/24/go-naming-tips.html

File: projects/aws/upgrader/upgrade/syscalls.go [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+	ReadFile(string) ([]byte, error)
+	OpenFile(string, int, os.FileMode) (*os.File, error)
+	Stat(string) (os.FileInfo, error)
+	Executable() (string, error)
+	ExecCommand(context.Context, string, ...string) ([]byte, error)
+	MkdirAll(string, os.FileMode) error
+}
+
+type sysCalls struct {
+	SysCalls
```
Comment (jacobweinstock) [inline projects/aws/upgrader/upgrade/syscalls.go:23]: This doesnt seem right. The empty struct already implements the interface, why embed the interface? Also, you dont need an interface at all.

**Review by jacobweinstock — CHANGES_REQUESTED** [review]

File: projects/aws/upgrader/Makefile (hunk 2 above)
Comment (jaxesn) [reply projects/aws/upgrader/Makefile:26]: yeah remove those two lines, 25 and 26

File: projects/aws/upgrader/Makefile [hunk 18]
```diff
@@ -34,6 +45,36 @@ PROJECT_DEPENDENCIES=eksd/kubernetes/client eksd/kubernetes/server eksd/cni-plug
 
 include $(BASE_DIRECTORY)/Common.mk
 
+TOOLS_BIN_DIR := $(shell pwd)/hack/tools/bin
```
Comment (jaxesn) [inline projects/aws/upgrader/Makefile:48]: move these 3 lines above the include with the rest of the "vars".  remove the spaces around the :=. instead of shell pwd use $(MAKE_ROOT) or something like that

File: projects/aws/upgrader/Makefile [hunk 19]
```diff
@@ -34,6 +45,36 @@ PROJECT_DEPENDENCIES=eksd/kubernetes/client eksd/kubernetes/server eksd/cni-plug
...
+MOCKGEN := $(TOOLS_BIN_DIR)/mockgen
+
+GO=build::common::use_go_version $(GOLANG_VERSION)
```
Comment (jaxesn) [inline projects/aws/upgrader/Makefile:51]: i like this idea, move above the include.  this feels like a pattern we could use more generally for cases where we call go in these makefiles to ensure we are always setting the go_version

File: projects/aws/upgrader/Makefile [hunk 20]
```diff
@@ -34,6 +45,36 @@ PROJECT_DEPENDENCIES=eksd/kubernetes/client eksd/kubernetes/server eksd/cni-plug
...
+	@git remote add origin https://github.com/aws/eks-anywhere-build-tooling.git 
+
+$(TOOLS_BIN_DIR):
+	@mkdir -p $(TOOLS_BIN_DIR)
+
+$(MOCKGEN): $(TOOLS_BIN_DIR)
+	GOBIN=$(TOOLS_BIN_DIR) $(GO) install github.com/golang/mock/mockgen@v1.6.0
+
+unit-test: | $$(ENABLE_DOCKER)
+	$(GO) test ./...
```
Comment (jaxesn) [inline projects/aws/upgrader/Makefile:73]: does this work properly, as in does it run thru docker?

**Review by jaxesn — CHANGES_REQUESTED** [review]
Just reviewed the makefile changes. look pretty decent!

File: projects/aws/upgrader/cmd/upgrade_node.go (hunk 3 above)
Comment (rahulbabu95) [reply projects/aws/upgrader/cmd/upgrade_node.go:40]: with flags we could have some basic validations but with args as string it's harder to validate. is that what you are asking.?

File: projects/aws/upgrader/upgrade/cniplugins_upgrade_test.go (hunk 7 above)
Comment (rahulbabu95) [reply projects/aws/upgrader/upgrade/cniplugins_upgrade_test.go:13]: All os/sys calls could return an error. And we aren't having any special error handling logic in the main flow. I don't see a value in adding/covering tests with an error scenario for each of the os calls. That would make the test file unreasonably big with little added value. Goal or idea here was to cover the happy flow to see the sequence of calls being made were validated and as expected.

File: projects/aws/upgrader/upgrade/syscalls.go (hunk 11 above)
Comment (rahulbabu95) [reply projects/aws/upgrader/upgrade/syscalls.go:12]: The idea of interfaces was to optimize for testability. We didn't want to make any actual system calls but also have a way to test the different upgrade flows. This was we could have the dependency injected and mocked out for out tests. I felt the size of interface was a reasonable trade-off given that these aren't super decoupled functions and all of them implicitly end up making os calls. I didn't feel the interface was necessarily bloated, but I will split it up in a more reasonable way if it adds value.

File: projects/aws/upgrader/upgrade/syscalls.go (hunk 17 above)
Comment (rahulbabu95) [reply projects/aws/upgrader/upgrade/syscalls.go:23]: The idea was to use an interface constructor and i need it for the testability reasons stated above. But i don't think i need to embed it though. I will clean up!

File: projects/aws/upgrader/Makefile (hunk 2 above)
Comment (rahulbabu95) [reply projects/aws/upgrader/Makefile:26]: Looks like we already had the `HAS_LICENSES=false` value set in our previous make. did something change between then and now to not have it .?

File: projects/aws/upgrader/cmd/root.go [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+func rootPersistentPreRun(cmd *cobra.Command, args []string) {
+	if err := initLogger(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func initLogger() error {
+	if err := logger.Init(viper.GetInt("verbosity")); err != nil {
```
Comment (abhinavmpandey08) [inline projects/aws/upgrader/cmd/root.go:34]: do we have logs with different verbosity?

File: projects/aws/upgrader/cmd/status.go [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+	"log"
+
+	"github.com/aws/eks-anywhere-build-tooling/projects/aws/upgrader/upgrade"
+	"github.com/spf13/cobra"
+)
+
+var upgradeStatusCmd = &cobra.Command{
+	Use:   "status",
+	Short: "Upgrade status",
+	Long:  "Use InPlace Upgrader upgrade status to get status of upgraded components on the node",
```
Comment (abhinavmpandey08) [inline projects/aws/upgrader/cmd/status.go:15]:
```suggestion
	Long:  "Use upgrade status command to get status of upgraded components on the node",
```

File: projects/aws/upgrader/cmd/upgrade_cniplugins.go [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+	"log"
+
+	"github.com/aws/eks-anywhere-build-tooling/projects/aws/upgrader/upgrade"
+	"github.com/spf13/cobra"
+)
+
+var upgradeCniPluginsCmd = &cobra.Command{
+	Use:   "cni-plugins",
+	Short: "Upgrade cni-plugins",
+	Long:  "Use InPlace Upgrader upgrade cni-plugins to upgrade cni-plugins on the node",
```
[the remark this answers is not in the document]
Comment (abhinavmpandey08) [reply projects/aws/upgrader/cmd/upgrade_cniplugins.go:15]:
```suggestion
	Long:  "Use upgrade cni-plugins command to upgrade cni-plugins on the node",
```

File: projects/aws/upgrader/cmd/upgrade_cniplugins.go [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+		}
+		return nil
+	},
+}
+
+func init() {
+	upgradeCmd.AddCommand(upgradeCniPluginsCmd)
+}
+
+func upgradeCniPlugins(ctx context.Context) error {
```
Comment (abhinavmpandey08) [inline projects/aws/upgrader/cmd/upgrade_cniplugins.go:29]: nit:
```suggestion
func upgradeCNIPlugins(ctx context.Context) error {
```

File: projects/aws/upgrader/cmd/upgrade_containerd.go [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+	"log"
+
+	"github.com/aws/eks-anywhere-build-tooling/projects/aws/upgrader/upgrade"
+	"github.com/spf13/cobra"
+)
+
+var upgradeContainerdCmd = &cobra.Command{
+	Use:   "containerd",
+	Short: "Upgrade containerd",
+	Long:  "Use InPlace Upgrader upgrade containerd to upgrade containerd on the node",
```
Comment (abhinavmpandey08) [inline projects/aws/upgrader/cmd/upgrade_containerd.go:15]:
```suggestion
	Long:  "Use upgrade containerd command to upgrade containerd on the node",
```

File: projects/aws/upgrader/cmd/upgrade_kubelet_kubectl.go [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+	"log"
+
+	"github.com/aws/eks-anywhere-build-tooling/projects/aws/upgrader/upgrade"
+	"github.com/spf13/cobra"
+)
+
+var upgradeKubeletKubectlCmd = &cobra.Command{
+	Use:   "kubelet-kubectl",
+	Short: "Upgrade kubelet-kubectl",
+	Long:  "Use InPlace Upgrader upgrade kubelet-kubectl to upgrade kubelet and kubectl on the node",
```
Comment (abhinavmpandey08) [inline projects/aws/upgrader/cmd/upgrade_kubelet_kubectl.go:15]:
```suggestion
	Long:  "Use upgrade kubelet-kubectl commamd to upgrade kubelet and kubectl on the node",
```

File: projects/aws/upgrader/cmd/upgrade_node.go [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+	Short: "Upgrade node",
+	Long:  "Use InPlace Upgrader upgrade node to upgrade kubeadm on the node",
```
Comment (abhinavmpandey08) [inline projects/aws/upgrader/cmd/upgrade_node.go:17]:
```suggestion
	Long:  "Use upgrade node command to upgrade kubeadm on the node",
```

File: projects/aws/upgrader/Makefile (hunk 2 above)
Comment (jaxesn) [reply projects/aws/upgrader/Makefile:26]: yes its a go bin now so its deps and therefore licenses and attribtion.

File: projects/aws/upgrader/cmd/root.go (hunk 21 above)
Comment (rahulbabu95) [reply projects/aws/upgrader/cmd/root.go:34]: No we don't at the moment. But this gives in the flexibility to add at different levels and limit the verbose at the level we want.!

File: projects/aws/upgrader/upgrade/kubeadm_upgrade.go [hunk 28]
```diff
@@ -0,0 +1,272 @@
...
+	kubeAdmUpgCmd := []string{"kubeadm", "upgrade", "apply", "--ignore-preflight-errors=CoreDNSUnsupportedPlugins,CoreDNSMigration", "--config", newKubeAdmConfig, "--allow-experimental-upgrades", "--yes"}
+	kubeAdmUpg, err := u.ExecCommand(ctx, kubeAdmUpgCmd[0], kubeAdmUpgCmd[1:]...)
+	if err != nil {
+		return execError(kubeAdmUpgCmd, string(kubeAdmUpg))
+	}
+	logger.Info("verbose output for kubeadm upgrade", "output", string(kubeAdmUpg))
+
+	upgCmpDir, err := u.upgradeComponentsDir()
+	if err != nil {
+		return fmt.Errorf("getting upgrade componenets directory: %v", err)
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubeadm_upgrade.go:83]:
```suggestion
		return fmt.Errorf("getting upgrade components directory: %v", err)
```

File: projects/aws/upgrader/upgrade/kubeadm_upgrade.go [hunk 29]
```diff
@@ -0,0 +1,272 @@
...
+	}
+	logger.Info("Upgraded kubeadm in First Control Plane successfully!", "version", u.kubernetesVersion)
+
+	return nil
+}
+
+func (u *Upgrader) KubeAdmInRestCP(ctx context.Context) error {
+	componentsDir, err := u.upgradeComponentsKubernetesBinDir()
+	if err != nil {
+		return fmt.Errorf("getting upgrade componenets kubernetes binary directory: %v", err)
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubeadm_upgrade.go:106]:
```suggestion
		return fmt.Errorf("getting upgrade components kubernetes binary directory: %v", err)
```

File: projects/aws/upgrader/upgrade/kubeadm_upgrade.go [hunk 30]
```diff
@@ -0,0 +1,272 @@
...
+	}
+
+	if err = u.BackUpAndReplace(kubeAdmBinDir, componentsDir, fmt.Sprintf("%s/kubeadm", componentsDir)); err != nil {
+		return fmt.Errorf("backing up and replacing kubeadm binary: %v", err)
+	}
+	logger.Info("Backed up and replaced kubeadm binary sucessfully")
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubeadm_upgrade.go:112]:
```suggestion
	logger.Info("Backed up and replaced kubeadm binary successfully")
```

File: projects/aws/upgrader/upgrade/kubeadm_upgrade.go [hunk 31]
```diff
@@ -0,0 +1,272 @@
...
+	kubeAdmUpgNodeCmd := []string{"kubeadm", "upgrade", "node", "--ignore-preflight-errors=CoreDNSUnsupportedPlugins,CoreDNSMigration"}
+	kubeAdmUpg, err := u.ExecCommand(ctx, kubeAdmUpgNodeCmd[0], kubeAdmUpgNodeCmd[1:]...)
+	if err != nil {
+		return execError(kubeAdmUpgNodeCmd, string(kubeAdmUpg))
+	}
+	logger.Info("verbose output for kubeadm upgrade", "output", string(kubeAdmUpg))
+
+	upgCmpDir, err := u.upgradeComponentsDir()
+	if err != nil {
+		return fmt.Errorf("getting upgrade componenets directory: %v", err)
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubeadm_upgrade.go:134]:
```suggestion
		return fmt.Errorf("getting upgrade components directory: %v", err)
```

File: projects/aws/upgrader/upgrade/kubeadm_upgrade.go [hunk 32]
```diff
@@ -0,0 +1,272 @@
...
+	}
+	logger.Info("Kubeadm in Control Plane upgraded successfully", "version", u.kubernetesVersion)
+
+	return nil
+}
+
+func (u *Upgrader) KubeAdmInWorker(ctx context.Context) error {
+	componentsDir, err := u.upgradeComponentsKubernetesBinDir()
+	if err != nil {
+		return fmt.Errorf("getting upgrade componenets kubernetes binary directory: %v", err)
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubeadm_upgrade.go:157]:
```suggestion
		return fmt.Errorf("getting upgrade components kubernetes binary directory: %v", err)
```

File: projects/aws/upgrader/upgrade/kubeadm_upgrade.go [hunk 33]
```diff
@@ -0,0 +1,272 @@
...
+	}
+
+	if err = u.BackUpAndReplace(kubeAdmBinDir, componentsDir, fmt.Sprintf("%s/kubeadm", componentsDir)); err != nil {
+		return fmt.Errorf("backing up and replacing kubeadm binary: %v", err)
+	}
+	logger.Info("Backed up and replaced kubeadm binary sucessfully")
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubeadm_upgrade.go:163]:
```suggestion
	logger.Info("Backed up and replaced kubeadm binary successfully")
```

File: projects/aws/upgrader/upgrade/kubelet_kubectl_upgrade.go [hunk 34]
```diff
@@ -0,0 +1,103 @@
...
+	kubeletBinPath                 = "/usr/bin/kubelet"
+	kubeletConf                    = "/etc/sysconfig/kubelet"
+	kubeletCredProviderFeatureGate = " --feature-gates=KubeletCredentialProviders=true"
+	extraArgs                      = "extra_args"
+)
+
+func (u *Upgrader) KubeletKubectlUpgrade(ctx context.Context) error {
+	cmpDir, err := u.upgradeComponentsKubernetesBinDir()
+	if err != nil {
+		return fmt.Errorf("getting upgrade componenets binary directory: %v", err)
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubelet_kubectl_upgrade.go:25]:
```suggestion
		return fmt.Errorf("getting upgrade components binary directory: %v", err)
```

File: projects/aws/upgrader/upgrade/kubelet_kubectl_upgrade.go [hunk 35]
```diff
@@ -0,0 +1,103 @@
...
+	stopKubeletCmd := []string{"systemctl", "stop", "kubelet"}
+	out, err := u.ExecCommand(ctx, stopKubeletCmd[0], stopKubeletCmd[1:]...)
+	if err != nil {
+		return execError(stopKubeletCmd, string(out))
+	}
+
+	if err := u.BackUpAndReplace(kubeletBinPath, cmpDir, fmt.Sprintf("%s/kubelet", cmpDir)); err != nil {
+		return fmt.Errorf("backing up and replacing kubelet binary: %v", err)
+	}
+	logger.Info("Backed up and replaced kubelet sucessfully")
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubelet_kubectl_upgrade.go:42]:
```suggestion
	logger.Info("Backed up and replaced kubelet successfully")
```

File: projects/aws/upgrader/upgrade/kubelet_kubectl_upgrade.go [hunk 36]
```diff
@@ -0,0 +1,103 @@
...
+	}
+
+	conf, err := u.ReadFile(kubeletConf)
+	if err != nil {
+		return fmt.Errorf("reading kubelet config on the node: %v", err)
+	}
+	newConf := bytes.ReplaceAll(conf, []byte(kubeletCredProviderFeatureGate), []byte(""))
+
+	extraArgsDir := fmt.Sprintf("%s/%s", cmpDir, extraArgs)
+	if err = u.MkdirAll(extraArgsDir, 0o755); err != nil {
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubelet_kubectl_upgrade.go:90]: Const these as well?

File: projects/aws/upgrader/upgrade/kubeadm_test.go [hunk 37]
```diff
@@ -0,0 +1,1003 @@
...
+	rotateCertificates: true
+	staticPodPath: /etc/kubernetes/manifests
+	`
+	upgCompBinDir = "/foo/binaries/kubernetes/usr/bin"
+	kubeVipBackup = "/foo/kube-vip.backup.yaml"
+	newKubeVip    = "/foo/kube-vip.yaml"
+	coreDNSBackup = "/foo/binaries/kubernetes/usr/bin/coredns.yaml"
+)
+
+func TestKubeAdmFistCPBackupExist(t *testing.T) {
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubeadm_test.go:78]: For all of them
```suggestion
func TestKubeAdmFirstCPBackupExist(t *testing.T) {
```

File: projects/aws/upgrader/upgrade/kubelet_kubectl.go [hunk 38]
```diff
@@ -0,0 +1,110 @@
...
+
+	extraArgsDir := fmt.Sprintf("%s/%s", cmpDir, extraArgs)
+	if err = u.MkdirAll(extraArgsDir, fileMode493); err != nil {
+		return fmt.Errorf("creating folder: %v", err)
+	}
+	kubeletConfBackupFile := fmt.Sprintf("%s/kubelet.bk", extraArgsDir)
+	if err = u.copy(kubeletConf, kubeletConfBackupFile); err != nil {
+		return copyError(kubeletConf, kubeletConfBackupFile, err)
+	}
+	if err := u.WriteFile(kubeletConf, newConf, fileMode416); err != nil {
```
Comment (vivek-koppuru) [inline projects/aws/upgrader/upgrade/kubelet_kubectl.go:104]: Why are these file modes called 493 and 416? I thought its like 755 or something

**Review by jacobweinstock — APPROVED** [review]

**Review by vivek-koppuru — APPROVED** [review]

File: projects/aws/upgrader/docker/linux/Dockerfile [hunk 39]
```diff
@@ -22,6 +22,7 @@ COPY _output/$RELEASE_BRANCH/dependencies/$TARGETOS-$TARGETARCH/eksd/kubernetes/
     _output/$RELEASE_BRANCH/dependencies/$TARGETOS-$TARGETARCH/eksd/kubernetes/server/bin/kubeadm \
     /eksa-upgrades/binaries/kubernetes/usr/bin/
 COPY ../../upgrade.sh /eksa-upgrades/scripts/
```
Comment (jaxesn) [inline projects/aws/upgrader/docker/linux/Dockerfile:24]: do we still need the sh script?

Comment (rahulbabu95) [reply projects/aws/upgrader/docker/linux/Dockerfile:24]: No we don't. but i was planning to remove the script in a separate PR once the eks-a [change](https://github.com/aws/eks-anywhere/pull/7927) for consuming the binary merges not to break CI. Given the nightly runs we shouldn't have problem but still ideally we want to remove once that is merged.

**Review by jaxesn — APPROVED** [review]

File: projects/aws/upgrader/LICENSE [hunk 40]
```diff
@@ -0,0 +1,175 @@
+
```
Comment (abhay-krishna) [inline projects/aws/upgrader/LICENSE:1]: Can you also generate and check in the ATTRIBUTION.txt file for this project, now that it's a Go project? You will also need to copy it into the image in the Dockerfile like you do the other attributions.

File: projects/aws/upgrader/upgrade/kubelet_kubectl.go (hunk 38 above)
Comment (abhay-krishna) [reply projects/aws/upgrader/upgrade/kubelet_kubectl.go]: 755 octal = 493 decimal
640 octal = 416 decimal

But I think almost nobody recognizes the decimal values for file permissions and is familiar with the octal values, and so we should definitely use those as variable names to avoid confusion.

File: projects/aws/upgrader/upgrade/cniplugins.go [hunk 41]
```diff
@@ -0,0 +1,37 @@
...
+
+// CNIPluginsUpgrade upgrades cni plugins on the node
+func (u *InPlaceUpgrader) CNIPluginsUpgrade(ctx context.Context) error {
+	cmpDir, err := u.upgradeComponentsBinDir()
+	if err != nil {
+		return fmt.Errorf("getting upgrade components binary directory: %v", err)
+	}
+
+	cniVersionCmd := []string{"/opt/cni/bin/loopback", "--version"}
+	out, err := u.ExecCommand(ctx, cniVersionCmd[0], cniVersionCmd[1:]...)
```
Comment (abhinavmpandey08) [inline projects/aws/upgrader/upgrade/cniplugins.go:18]: Why is this needed and where is it being used?

File: projects/aws/upgrader/upgrade/containerd.go [hunk 42]
```diff
@@ -0,0 +1,52 @@
...
+//  1. copy over the new containerd components
+//  2. reload the daemon and restart containerd with the new version
+func (u *InPlaceUpgrader) ContainerdUpgrade(ctx context.Context) error {
+	cmpDir, err := u.upgradeComponentsBinDir()
+	if err != nil {
+		return fmt.Errorf("getting upgrade components binary directory: %v", err)
+	}
+
+	containerdVersionCmd := []string{"containerd", "--version"}
+	out, err := u.ExecCommand(ctx, containerdVersionCmd[0], containerdVersionCmd[1:]...)
```
Comment (abhinavmpandey08) [inline projects/aws/upgrader/upgrade/containerd.go:21]: Same question here. Let’s at least log the before version if we are calling it

**Review by abhinavmpandey08 — APPROVED** [review]
/approve
/lgtm
/woof
